### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by FinHardy**
Stories 7,8,9 and 10

### CODE REVIEW

File: src/components/BookPage/AddReview/index.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import { useEffect, useState } from "react";
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: useEffect isn't used

File: src/components/BookPage/AddReview/index.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import "./addreview.css";
+
+function Form(props) {
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: The name form is way to generic. Call it something like AddReviewForm

File: src/components/BookPage/index.js [hunk 3]
```diff
@@ -76,6 +77,7 @@ function BookPage() {
          
 
           <p>{blurb}</p>
+          <AddReview id={id}/>
```
Comment (ashleycoles) [inline src/components/BookPage/index.js]: I should be able to see the new review as soon as I add it.

File: src/components/BookPage/AddReview/index.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      isValid = false;
+    }
+
+    if (isValid) {
+      fetch("https://book-swap-api.dev.io-academy.uk/api/reviews", {
+        method: "POST",
+        body: JSON.stringify({
+          name: `${name}`,
+          rating: rating,
+          review: `${review}`,
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: Why use template literals here?

Just do 

{
    name: name,
   ......
}

File: src/components/BookPage/AddReview/index.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          Accept: "application/json",
+        },
+      })
+        .then((res) => res.json())
+        .then((data) => {
+          console.log(data);
+          setName("");
+          setRating("");
+          setReview("");
+          setReviewCharCount(0);
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: wat. What about errors that come back from the API? 

I should also be able to see the new review immediately

File: src/components/BookPage/AddReview/index.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        });
+    }
+  }
+
+  function changeName(event) {
+    setName(event.target.value);
+  }
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: Can you move this up to the other change handlers, and make sure they are named consistently

Comment (ashleycoles) [reply src/components/BookPage/AddReview/index.js]: Or make it a fat arrow anon callback like you have with with the rating field

File: src/components/BookPage/AddReview/index.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      />
+      <div className="error-message">{ratingError}</div>
+
+      <label className="label-names" htmlFor="review">
+        Review
+      </label>
+      <textarea id="review" value={review} onChange={handleReviewChange} />
+      <div className="char-count">
+        {reviewCharCount} / {maxReviewLength} characters
+      </div>
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: I like this, nice addition

File: src/components/BookPage/AddReview/index.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+          setReviewCharCount(0);
+        });
+    }
+  }
+
+  function changeName(event) {
+    setName(event.target.value);
+  }
+
+  const id = props.id;
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: Why is this here? It only gets used in the fetch request to post the review, so move it into the function that uses it

File: src/components/BookPage/AddReview/index.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+      setRatingError("Rating field is required");
+      isValid = false;
+    }
+
+    if (review.trim() === "") {
+      setReviewError("Review field is required");
+      isValid = false;
+    }
+
+    if (isValid) {
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: I like this validation logic, but this handleSubmit function is now doing multiple things, validation and submission. 

You should split the validation into it's own function

File: src/components/BookPage/AddReview/index.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+  function changeName(event) {
+    setName(event.target.value);
+  }
+
+  const id = props.id;
+
+  return (
+    <form className="review-form" onSubmit={handleSubmit}>
+      <p className="review-title">Want to review this books?</p>
```
Comment (ashleycoles) [inline src/components/BookPage/AddReview/index.js]: Extra s after book

File: src/components/GenreFilter/index.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  function setGenre(event) {
+    props.setGenreID(event.target.value);
+  }
+
+  function setGenre(event) {
+    if (props.setGenreID) {
+      props.setGenreID(event.target.value);
+    } else if (props.ClaimedGenreID) {
+      props.setClaimedGenreID(event.target.value);
+    }
```
Comment (ashleycoles) [inline src/components/GenreFilter/index.js]: hmmm I don't like this. Ultimately setting a genre is really the same if it's claimed or not. The necessity for this if statement tells me that your approach to splitting out claimed/unclaimed might not be working out so well now.

File: src/components/HomePage/index.js [hunk 12]
```diff
@@ -1,27 +1,48 @@
...
-import { Link } from 'react-router-dom'; 
-import BookCard from '../BookCard';
+import React, { useEffect, useState } from "react";
+import { Link } from "react-router-dom";
+import BookCard from "../BookCard";
+import GenreFilter from "../GenreFilter";
+import "./home.css";
+import SearchBar from "../SearchBar";
 
 function HomePage() {
```
Comment (ashleycoles) [inline src/components/HomePage/index.js]: There is a dryness issue with this component and the ClaimedBookPage. Lots of duplication, and it has added complexity in the way you're having to search and filter.

Since the only difference between this and the claimed page is one number (0 and 1 in the fetch), at this point you should refactor this into a single component that works for both claimed and unclaimed.

That will help simplify filtering and searching.

File: src/components/HomePage/index.js [hunk 13]
```diff
@@ -1,27 +1,48 @@
...
+  }, [genreID, search]);
+
+  return (
+    <div>
+      <div className="search-container">
+        <GenreFilter genreID={genreID} setGenreID={setGenreID} />
+        <SearchBar search={search} setSearch={setSearch} />
+      </div>
+      <div className="flex_container">
+        {books.map((book, index) => (
```
Comment (ashleycoles) [inline src/components/HomePage/index.js:40]: Worth now thinking about what happens if there are no books, which is possible with claimed/unclaimed. Just displaying a simple message if books is empty would do.

File: src/components/HomePage/index.js (hunk 12 above)
Comment (ashleycoles) [reply src/components/HomePage/index.js]: DO THIS FIRST

File: src/components/ClaimedBookPage/index.js [hunk 14]
```diff
@@ -1,27 +1,50 @@
-import React, { useEffect, useState } from 'react';
-import { Link } from 'react-router-dom'; // Import the Link component
-import BookCard from '../BookCard';
+import React, { useEffect, useState } from "react";
+import { Link } from "react-router-dom"; // Import the Link component
+import BookCard from "../BookCard";
+import GenreFilter from "../GenreFilter";
+import SearchBar from "../SearchBar";
 
 function ClaimedBookPage() {
```
Comment (ashleycoles) [inline src/components/ClaimedBookPage/index.js]: Please see my comments on the Homepage regarding DRYNESS

File: src/components/Nav/AddBook/index.js [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+      postData.image = imgUrl;
+    }
+
+    if (year != 0) {
+      postData.year = year;
+    }
+
+    if (pageCount != 0) {
+      postData.page_count = pageCount;
+    }
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: I would consider moving this into a function `generatePostData` or something just to keep each function as small as possible

File: src/components/Nav/AddBook/index.js [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+        "Accept": "application/json",
+      },
+    })
+      .then((res) => res.json())
+      .then((data) => {
+        if (data.errors) {
+          setTitleErrors(data.errors.title);
+          setAuthorErrors(data.errors.author);
+          setGenreErrors(data.errors.genre_id);
+          console.log(data);
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: Make sure to remove logs

File: src/components/Nav/AddBook/index.js [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+        }
+        setRedirect(true);
+      });
+
+    console.log("sucessfully submitted book!");
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: Remove

File: src/components/Nav/AddBook/index.js [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+    })
+      .then((res) => res.json())
+      .then((data) => {
+        if (data.errors) {
+          setTitleErrors(data.errors.title);
+          setAuthorErrors(data.errors.author);
+          setGenreErrors(data.errors.genre_id);
+          console.log(data);
+        }
+        setRedirect(true);
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: Redirect even if there are errors?

File: src/components/Nav/AddBook/index.js [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+          type="input"
+          name="genre"
+          placeholder="None selected"
+          onChange={changeGenre}
+        >
+          <option value="">Select an option</option>
+          <option value="1">Thrlller</option>
+          <option value="2">Romance</option>
+          <option value="3">Historical</option>
+          <option value="4">Non-fiction</option>
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: What if the list of genres changes in the API? You should be dynamically getting the genres here. Never hardcode data that comes from an API

File: src/components/Nav/AddBook/index.js [hunk 20]
```diff
@@ -0,0 +1,177 @@
+import { useState } from "react";
+import "./add-book.css";
+import { Navigate } from "react-router-dom";
+
+function AddBook() {
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: Why is this inside Nav?

File: src/components/Nav/AddBook/index.js [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+          onChange={changeImgUrl}
+        />
+      </div>
+      <div className="field">
+        <textarea
+          placeholder="Blurb"
+          rows="4"
+          cols="50"
+          onChange={changeBlurb}
+        ></textarea>
```
Comment (ashleycoles) [inline src/components/Nav/AddBook/index.js]: This input is missing a label

**Review by ashleycoles — CHANGES_REQUESTED** [review]
Some changes to keep you busy!

File: src/components/BookPage/AddReview/index.js (hunk 1 above)
Comment (mattdyer1984) [reply src/components/BookPage/AddReview/index.js]: done

File: src/components/BookPage/AddReview/index.js (hunk 2 above)
Comment (mattdyer1984) [reply src/components/BookPage/AddReview/index.js]: done

File: src/components/BookPage/index.js (hunk 3 above)
Comment (mattdyer1984) [reply src/components/BookPage/index.js]: done

File: src/components/BookPage/AddReview/index.js (hunk 5 above)
Comment (mattdyer1984) [reply src/components/BookPage/AddReview/index.js]: done

File: src/components/BookPage/AddReview/index.js (hunk 10 above)
Comment (mattdyer1984) [reply src/components/BookPage/AddReview/index.js]: done

File: src/components/SearchBar/index.js [hunk 22]
```diff
@@ -15,7 +15,7 @@ function SearchBar(props) {
 
   return (
     <div className="searchbar">
-      <FaSearch id="search-icon" onClick={handleSubmit} />
+      <FaSearch id="search-icon" onClick={handleSubmit}/>
```
Comment (ashleycoles) [inline src/components/SearchBar/index.js]: Actually, I don't think it's very obvious that clicking the icon is how you submit the search. I would suggest either of the following approaches:

Use onChange on the input itself so it searches as you type

Add a search button

File: src/App.js [hunk 23]
```diff
@@ -3,19 +3,23 @@ import "./App.css";
...
-import ClaimedBookPage from "./components/ClaimedBookPage";
-import AddBook from "./components/Nav/AddBook";
+// import ClaimedBookPage from "./components/ClaimedBookPage";
+import AddBook from "./components/AddBook";
+import { useState } from "react";
 
 function App() {
+  const [claimedUrl, setClaimedUrl] = useState(
+    "https://book-swap-api.dev.io-academy.uk/api/books?claimed=0"
+  );
```
Comment (ashleycoles) [inline src/App.js]: Storing the entire URL here is overkill, the only thing that needs to change for each route is 0 and 1, the rest of the URL remains the same. It would be better to go for a simpler solution

File: src/components/Nav/index.js [hunk 24]
```diff
@@ -1,13 +1,29 @@
...
+    props.setClaimedUrl(
+      "https://book-swap-api.dev.io-academy.uk/api/books?claimed=0"
+    );
+  }
+
+  function handleClaimedClick() {
+    props.setClaimedUrl(
+      "https://book-swap-api.dev.io-academy.uk/api/books?claimed=1"
+    );
+  }
```
Comment (ashleycoles) [inline src/components/Nav/index.js]: While this approach works, it's definitely overkill

File: src/components/Nav/index.js [hunk 25]
```diff
@@ -1,13 +1,29 @@
...
+      <a href="/" className="bookSwap">Books Swap</a>
       <div className="right-links">
-        <Link to="/">Available books</Link>
-        <Link to="/claimed/">Claimed books</Link>
+        <Link to="/" onClick={handleAvailableClick}>
+          Available books
+        </Link>
+        <Link to="/" onClick={handleClaimedClick}>
+          Claimed Book
+        </Link>
```
Comment (ashleycoles) [inline src/components/Nav/index.js]: Now it's not possible to link to claimed/unclaimed books. It was better when you had two separate pages / and /claimed

File: src/App.js [hunk 26]
```diff
@@ -3,19 +3,23 @@ import "./App.css";
...
     <div className="App">
       <BrowserRouter>
-        <Nav />
+        <Nav setClaimedUrl={setClaimedUrl} claimedUrl={claimedUrl} />
 
         <Routes>
           <Route path="/book/:id" element={<BookPage />} />
-          <Route path="/" element={<HomePage />} />
-          <Route path="/claimed/" element={<ClaimedBookPage />} />
+          <Route path="/" element={<HomePage claimedUrl={claimedUrl} />} />
```
Comment (ashleycoles) [inline src/App.js]: You should still have two routes for / and claimed.

Rather than passing in the entire URL using this system shared between app and the Nav, it's better to just have two routes that both display the HomePage (you should rename it though, it's not just the homepage anymore) but pass in a prop that sets whether or not you want claimed books.

File: src/components/AddBook/index.js [hunk 27]
```diff
@@ -118,11 +131,14 @@ function AddBook() {
...
-          <option value="4">Non-fiction</option>
+          <option className="dropdown" value={0}>
+            All
+          </option>
+          {bookGenres.map((genre) => (
+            <option className="dropdown" value={genre.id} key={genre.id}>
+              {genre.name}
+            </option>
+          ))}
         </select>
```
Comment (ashleycoles) [inline src/components/AddBook/index.js]: I would suggest moving this into it's own component, along with the actual fetching of the genres. Trying to keep each component as small as possible.

File: src/components/AddBook/index.js [hunk 28]
```diff
@@ -82,12 +97,10 @@ function AddBook() {
           setTitleErrors(data.errors.title);
           setAuthorErrors(data.errors.author);
           setGenreErrors(data.errors.genre_id);
-          console.log(data);
+        } else {
+          setRedirect(true);
```
Comment (ashleycoles) [inline src/components/AddBook/index.js:81]: I've just noticed there are some other API errors you might get that you haven't accounted for:

blurb
image
year

These three also have API validation the same as title, author and genre_id, so worth adding error messages for them exactly like you have already for title, author and genre_id.

To see what I mean, try and add a book with a description longer than 500 chars, an image URL that isn't a URL and a year above 2100

File: src/components/BookPage/AddReviewForm/index.js [hunk 29]
```diff
@@ -11,6 +11,11 @@ function Form(props) {
   const [reviewCharCount, setReviewCharCount] = useState(0);
   const maxReviewLength = 450;
 
+  function handleInputChange(event, setState) {
+    const newValue = event.target.value;
+    setState(newValue);
+  }
```
Comment (ashleycoles) [inline src/components/BookPage/AddReviewForm/index.js:17]: This is a nice approach. Please do this on the AddBook page as well for consistency

File: src/components/AddBook/index.js [hunk 30]
```diff
@@ -11,6 +11,8 @@ function AddBook() {
   const [imgUrl, setImgUrl] = useState("");
   const [blurb, setBlurb] = useState("");
 
+  const [bookGenres, setBookGenres] = useState([]);
+
   const [redirect, setRedirect] = useState(false);
 
   const [titleErrors, setTitleErrors] = useState("");
```
Comment (ashleycoles) [inline src/components/AddBook/index.js]: (wont let me comment on the correct lines, but please see my comment on AddReviewForm about your cool change handler, please apply that approach here as well

File: src/components/BookPage/AddReviewForm/index.js [hunk 31]
```diff
@@ -21,63 +26,47 @@ function Form(props) {
...
+          setRatingError(data.errors.rating);
+          setReviewError(data.errors.review);
+        } else {
+          setNameError("");
+          setRatingError("");
+          setReviewError("");
           setName("");
           setRating("");
           setReview("");
           setReviewCharCount(0);
```
Comment (ashleycoles) [inline src/components/BookPage/AddReviewForm/index.js]: This is getting a bit chunky now, you should move all these into a function 'clearForm' or something

File: src/components/BookPage/AddReviewForm/index.js [hunk 32]
```diff
@@ -21,63 +26,47 @@ function Form(props) {
...
-        });
-    }
+          props.onReviewSubmit({
+            name: name,
+            rating: rating,
+            review: review,
+          });
```
Comment (ashleycoles) [inline src/components/BookPage/AddReviewForm/index.js:67]: A better approach would be to instead get the BookPage to re-fetch the book when you submit the review form.

File: src/components/BookPage/index.js [hunk 33]
```diff
@@ -48,6 +48,8 @@ function BookPage() {
       })
       .catch((error) => console.error("Error fetching data", error));
   }, [id]);
+  function handleReviewSubmit(newReview) {
+    setReviews([...reviews, newReview])}
```
Comment (ashleycoles) [inline src/components/BookPage/index.js]: I see why you're doing this, but a better way to do it would be to get handleReviewSubmit to just run the fetch request that's in your useEffect again. Just move that fetch request into it's own function that gets called in the useEffect, and gets called on review form submission instead of handleReviewSubmit.

Better for two reasons:

1) It's a bit simpler (Arguable)
2) It makes sure the data is always accurate

File: src/components/HomePage/home 2.css [hunk 34]
```diff
@@ -0,0 +1,9 @@
+.search-container {
```
Comment (ashleycoles) [inline src/components/HomePage/home 2.css]: home 2.css?

File: src/components/HomePage/index.js [hunk 35]
```diff
@@ -5,13 +5,13 @@ import GenreFilter from "../GenreFilter";
...
 
-function HomePage() {
+function HomePage(props) {
   const [books, setBooks] = useState([]);
   const [genreID, setGenreID] = useState(0);
   const [search, setSearch] = useState("");
 
   useEffect(() => {
-    let url = "https://book-swap-api.dev.io-academy.uk/api/books?claimed=0";
+    let url = props.claimedUrl;
```
Comment (ashleycoles) [inline src/components/HomePage/index.js]: See other comments about refactoring the way this works (in app and nav)

**Review by ashleycoles — CHANGES_REQUESTED** [review]
A few more changes for you!

File: src/components/BookPage/index.js (hunk 33 above)
Comment (mattdyer1984) [reply src/components/BookPage/index.js]: done

File: src/App.js [hunk 36]
```diff
@@ -19,7 +17,9 @@ function App() {
 
         <Routes>
           <Route path="/book/:id" element={<BookPage />} />
-          <Route path="/" element={<HomePage claimedUrl={claimedUrl} />} />
+          <Route path="/" element={<HomePage claimedUrl={0} />} />
+          <Route path="/available/" element={<HomePage claimedUrl={0} />} />
```
Comment (ashleycoles) [inline src/App.js:21]: I wouldn't bother with this route, just have home as unclaimed and /claimed as claimed (make sure to update the nav accordingly)

File: src/App.js [hunk 37]
```diff
@@ -8,9 +8,7 @@ import AddBook from "./components/AddBook";
 import { useState } from "react";
 
 function App() {
-  const [claimedUrl, setClaimedUrl] = useState(
-    "https://book-swap-api.dev.io-academy.uk/api/books?claimed=0"
-  );
+  const [claimedUrl, setClaimedUrl] = useState(0);
```
Comment (ashleycoles) [inline src/App.js:11]: You're not actually using this anymore. Get rid of it and remove it from the nav as well

File: src/components/AddBook/GenreDropdown/index.js [hunk 38]
```diff
@@ -0,0 +1,43 @@
...
+    const newValue = event.target.value;
+    setState(newValue);
+  }
+
+  useEffect(() => {
+    fetch("https://book-swap-api.dev.io-academy.uk/api/genres")
+      .then((res) => res.json())
+      .then((data) => {
+        setBookGenres(data.data);
+        console.log(data.data);
```
Comment (ashleycoles) [inline src/components/AddBook/GenreDropdown/index.js:17]: Remove logs

File: src/components/AddBook/index.js [hunk 39]
```diff
@@ -185,6 +150,9 @@ function AddBook() {
         <p>{titleErrors}</p>
         <p>{authorErrors}</p>
         <p>{genreErrors}</p>
```
Comment (ashleycoles) [inline src/components/AddBook/index.js:153]: Just noticed that the error you get is 'genre_id field is required' - that's going to confuse the hell out of a user. Because the API is dealing with IDs, but the user isn't.

File: src/components/BookPage/AddReviewForm/index.js [hunk 40]
```diff
@@ -16,6 +16,16 @@ function AddReviewForm(props) {
...
 
+  function clearForm() {
+          setNameError("");
+          setRatingError("");
+          setReviewError("");
+          setName("");
+          setRating("");
+          setReview("");
+          setReviewCharCount(0);
+
```
Comment (ashleycoles) [inline src/components/BookPage/AddReviewForm/index.js]: Remove empty line

File: src/components/BookPage/index.js [hunk 41]
```diff
@@ -47,12 +47,17 @@ function BookPage() {
         console.log(info.data.reviews.length);
       })
       .catch((error) => console.error("Error fetching data", error));
-  }, [id]);
+  }
+  useEffect(() => {
+    fetchBookData()
+  }, [id])
```
Comment (ashleycoles) [inline src/components/BookPage/index.js:53]: Empty line needed under this useffect

Comment (ashleycoles) [inline src/components/BookPage/index.js:53]: Now you can just pass fetchBookData directly into useffect

`useEffect(fetchBookData, [id])`

File: src/components/BookPage/index.js [hunk 42]
```diff
@@ -47,12 +47,17 @@ function BookPage() {
...
   function handleReviewSubmit(newReview) {
-    setReviews([...reviews, newReview])}
+    setReviews([...reviews, newReview])
```
Comment (ashleycoles) [inline src/components/BookPage/index.js:55]: You don't need to do this anymore. Because the form adds the new review, and then you re-fetch the book, you naturally get the new review.

No need to setReviews, or to pass the review into this function.

I would actually remove this function entirely, and rather than passing this down to the review form, instead pass down fetchBookData and call that upon submission

File: src/components/HomePage/index.js (hunk 13 above)
Comment (ashleycoles) [reply src/components/HomePage/index.js:40]: This comment still stands. Show a basic 'no books found' message if books is empty

**Review by ashleycoles — CHANGES_REQUESTED** [review]
A few more

**Review by ashleycoles — APPROVED** [review]
